agreement/asyncVoteVerifier.go

```diff
@@ -140,18 +118,14 @@ func (avv *AsyncVoteVerifier) verifyVote(verctx context.Context, l LedgerReader,
 		// if we're done while waiting for room in the requests channel, don't queue the request
 		req := asyncVerifyVoteRequest{ctx: verctx, l: l, uv: &uv, index: index, message: message, out: out}
 		avv.wg.Add(1)
-		if err := avv.backlogExecPool.EnqueueBacklog(avv.ctx, avv.executeVoteVerification, req, avv.execpoolOut); err != nil {
-			// we want to call "wg.Done()" here to "fix" the accounting of the number of pending tasks.
-			// if we got a non-nil, it means that our context has expired, which means that we won't see this task
-			// getting to the verification function.
-			avv.wg.Done()
-			return err
-		}
+		avv.batchInputChan <- &req
 	}
-	return nil
 }
 
-func (avv *AsyncVoteVerifier) verifyEqVote(verctx context.Context, l LedgerReader, uev unauthenticatedEquivocationVote, index uint64, message message, out chan<- asyncVerifyVoteResponse) error {
+func (avv *AsyncVoteVerifier) verifyEqVote(verctx context.Context, l LedgerReader, uev unauthenticatedEquivocationVote,
+	index uint64, message message, out chan<- asyncVerifyVoteResponse) {
+	avv.mu.RLock()
```

Comment:
Why do you need to take the lock here?
I think it looks safe without it, but perhaps I'm missing something.

Reply:
This issue was a long standing bug and this lock is fixing it. 
The comment on lines 130-132 gives some context (`case <-avv.ctx.Done(): // if we're quitting, don't ...`.

The `avv.wg` will wait for any pending votes in verification before shutting down `avv.backlogExecPool`. But in the event when `verifyEqVote` and `Quit` are called simultaneously, we can have this execution order which leads to a vote get into verification when the verifier is shut down, then the `avv.wg` will wait forever. 
`verifyEqVote`: `case <-avv.ctx.Done():`
`                                  Quit`: `avv.ctxCancel()`
`                                  Quit`: `avv.batchVerifier.WaitForStop()`
`verifyEqVote`: `avv.wg.Add(1)`
`verifyEqVote`: `avv.batchInputChan <- &req`
`                                  Quit`: `avv.wg.Wait()`

This is even worse in Master, where the output channel will be closed before the execpool sends the result through it, causing a panic. I have a test that reproduces the panic, and the fix is similar to this. See it in this PR [here](https://github.com/algorand/go-algorand/pull/5341).

I don't think the read lock is a considerable cost here. There could possibly be a more complex fix without a lock here, but I think it is not worth the complexity.

Reply:
I see. Okay, this is probably also the answer to my previous comment about a similar issue.